sympy/matrices/tests/test_matrices.py

```diff
@@ -330,6 +330,65 @@ def test_creation():
     raises(ValueError, lambda: Matrix([ones(2), ones(3, 0)]))
 
 
+def test_creation_16723():
+    # Tests for callback input and output sanitization
+    assert Matrix(
+            2, 2, lambda i, j: i + j,
+            sympify_callback_arguments=True,
+            sympify_callback_return=True
+            ).tolist() == \
+                [[S(0), S(1)], [S(1), S(2)]]
+
+    assert Matrix(
+            2, 2, lambda i, j: i + j,
+            sympify_callback_arguments=True,
+            sympify_callback_return=False
+            ).tolist() == \
+                [[S(0), S(1)], [S(1), S(2)]]
+
+    assert Matrix(
+            2, 2, lambda i, j: i + j,
+            sympify_callback_arguments=False,
+            sympify_callback_return=True
+            ).tolist() == \
+                [[S(0), S(1)], [S(1), S(2)]]
+
+    assert Matrix(
```

Comment:
```python
from sympy.core.basic import _aresame
assert _aresame(Matrix(2, 2, lambda i, j: i + j,
            sympify_callback_arguments=False,
            sympify_callback_return=False
            ).tolist(),
                [[0, 1], [1, 2]])
```

The `==` will not affirm that these values are not Integers. You could use it on all these assertions.